app/validators/answers/answer_validator.py

```diff
@@ -3,9 +3,40 @@
 
 
 class AnswerValidator(Validator):
-    def __init__(self, schema_element):
+    OPTION_MISSING_Q_CODE = "Option q_code must be provided"
+    ANSWER_MISSING_Q_CODE = "Answer q_code must be provided"
+    DETAIL_ANSWER_MISSING_Q_CODE = "Detail answer q_code must be provided"
+
+    def __init__(self, schema_element, questionnaire_schema):
         super().__init__(schema_element)
         self.answer = schema_element
         self.answer_id = self.answer["id"]
         self.answer_type = AnswerType(self.answer["type"])
         self.context["answer_id"] = self.answer_id
+        self.questionnaire_schema = questionnaire_schema
+
+    def validate(self):
+        if self.questionnaire_schema.schema["data_version"] == "0.0.1":
+            self._validate_q_codes()
+
+        return self.errors
+
+    def _validate_q_codes(self):
+        if not self.answer.get("q_code"):
```

Comment:
This is only ever checking if the answer doesn't have a q_code. But if answer had a q_code, but options also did, then it is invalid, or if detail answer then didn't have it for Radio, it wouldn't catch that either.

Reply:
Ok, refactored now, hope this is the logic we want.